mdata/cache/ccache_metric.go

```diff
@@ -69,6 +69,66 @@ func (mc *CCacheMetric) Del(ts uint32) int {
 	return len(mc.chunks)
 }
 
+func (mc *CCacheMetric) AddSlice(prev uint32, itergens []chunk.IterGen) {
+	if len(itergens) == 0 {
+		return
+	}
+
+	mc.Lock()
+	defer mc.Unlock()
+
+	ts := itergens[0].Ts
+
+	// if previous chunk has not been passed we try to be smart and figure it out.
+	// this is common in a scenario where a metric continuously gets queried
+	// for a range that starts less than one chunkspan before now().
+	if prev == 0 {
+		res, ok := mc.seekDesc(ts - 1)
+		if ok {
+			prev = res
+		}
+	}
+
+	for _, itergen := range itergens {
+		ts = itergen.Ts
+
+		if _, ok := mc.chunks[ts]; ok {
+			// chunk is already present. no need to error on that, just ignore it
+			continue
+		}
+
+		mc.chunks[ts] = &CCacheChunk{
+			Ts:    ts,
+			Prev:  0,
+			Next:  0,
+			Itgen: itergen,
+		}
+
+		// if the previous chunk is cached, link in both directions
+		if _, ok := mc.chunks[prev]; ok {
```

Comment:
in all but the first case, we know there was a prev
we can do most of the inserts such that we immediately set the right Prev and Next fields

Reply:
I think we can also pre-allocate the `&CCacheChunk` in a slice of len = len(itergens) for two benefits:
1) 1 allocation instead of N (faster Adds)
2) at read time, we're more likely to have a sequential read (prefetched from RAM)

obviously, needs benchmarking to see if this hunch is correct.